src/main/scala/utility/ClockGatedReg.scala

```diff
@@ -65,3 +66,79 @@ object GatedRegNextN {
     }
   }
 }
+
+class SegmentedAddr(_segments: Seq[Int]) {
+
+  def this(f: Parameters => Seq[Int])(implicit p: Parameters) = this(f(p))
+
+  val segments = _segments // (High, Lower ...)
+  private var addr = UInt(segments.sum.W)
+
+  // [High, Lower ...]
+  private def segment(addr_in: UInt): Seq[UInt] = {
+    segments.foldLeft((Seq[UInt](), addr_in.asBools)) { (acc, segment_length) =>
+      ((acc._1 :+ VecInit(acc._2.takeRight(segment_length)).asUInt), acc._2.dropRight(segment_length))
+    }._1
+  }
+
+  def getAddr(): UInt = {
+    addr
+  }
+  def getAddrSegments(): Seq[UInt] = {
+    segment(addr)
+  }
+  def fromSegments(seg: Seq[UInt]) = {
+    this.addr = seg.reduce(Cat(_, _))
+    this
+  }
+  def fromAddr(addr_in: UInt) = {
+    this.addr = addr_in
+    this
+  }
+
+  def compare(that: SegmentedAddr): Seq[Bool] = {
+    this.getAddrSegments() zip that.getAddrSegments() map {
+      case (l, r) => l =/= r
+    }
+  }
+
+  def compare(that: Seq[UInt]): Seq[Bool] = {
+    this.getAddrSegments() zip that map {
+      case (l, r) => l =/= r
+    }
+  }
+}
+
+object SegmentedAddrNext {
+  def apply(addr: SegmentedAddr): SegmentedAddr = {
+    apply(addr, true.B, None)
+  }
+
+  def apply(addr: SegmentedAddr, fire: Bool, parentName: Option[String]): SegmentedAddr = {
+    apply(addr.getAddr(), addr.segments, fire, parentName)
+  }
+
+  def apply(addr: UInt, segments: Seq[Int], fire: Bool, parentName: Option[String]): SegmentedAddr = {
+    // Input wire, segmented
+    val segmented = new SegmentedAddr(segments).fromAddr(addr).getAddrSegments()
+
+    val modified = Wire(Vec(segmented.length, Bool()))
+
+    val segmentedNext = segmented zip modified.zipWithIndex map {
+      case (seg, (modified, idx)) => RegEnable(seg, modified && fire)
+        .suggestName(s"${parentName.getOrElse("")}_seg_${idx}_value")
+    }
+    modified zip segmentedNext zip segmented map {
```

Comment:
It seems that the low bits still need to be compared, so where is the benefit of`SegmentedAddrNext`?

Reply:
This is just like GatedRegNext, but for convenience on PC high-low management.

```
  val s1_pc_dup   = s0_pc_dup.zip(io.s0_fire).map {case (s0_pc, s0_fire) => RegEnable(s0_pc, s0_fire)}
  val s2_pc_dup   = s1_pc_dup.zip(io.s1_fire).map {case (s1_pc, s1_fire) => SegmentedAddrNext(s1_pc, pcSegments, s1_fire, Some("s2_pc"))}
  val s3_pc_dup   = s2_pc_dup.zip(io.s2_fire).map {case (s2_pc, s2_fire) => SegmentedAddrNext(s2_pc, s2_fire, Some("s3_pc"))}
```

Reply:
I got, thanks.